src/platform/test/mtest-main.c

```diff
@@ -0,0 +1,206 @@
+/* Copyright (c) 2022 Felix Jones
+*
+* This Source Code Form is subject to the terms of the Mozilla Public
+* License, v. 2.0. If a copy of the MPL was not distributed with this
+* file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+#include <mgba/core/cheats.h>
+#include <mgba/core/config.h>
+#include <mgba/core/core.h>
+#include <mgba/core/serialize.h>
+#include <mgba/internal/gba/gba.h>
+
+#include <mgba/feature/commandline.h>
+#include <mgba-util/vfs.h>
+
+#include <signal.h>
+
+#define MTEST_OPTIONS "S:R:"
+#define MTEST_USAGE \
+	"\nAdditional options:\n" \
+	"  -S SWI           Run until specified SWI call before exiting\n" \
+	"  -R REGISTER      General purpose register to return as exit code\n" \
+
+struct MTestOpts {
+	int exitSwiImmediate;
+	int returnCodeRegister;
+};
+
+static void _mTestShutdown(int signal);
+static bool _parseMTestOpts(struct mSubParser* parser, int option, const char* arg);
+static int _parseNamedRegister(const char* regStr);
+
+static bool _dispatchExiting = false;
+static int _exitCode = 0;
+
+#ifdef M_CORE_GBA
+static void _mTestSwi3Callback(struct mCore* core);
+
+static int _exitSwiImmediate = -1;
+static int _returnCodeRegister;
+
+void (*_armSwi16)(struct ARMCore* cpu, int immediate);
+void (*_armSwi32)(struct ARMCore* cpu, int immediate);
+
+static void _mTestSwi16(struct ARMCore* cpu, int immediate);
+static void _mTestSwi32(struct ARMCore* cpu, int immediate);
+#endif
+
+int main(int argc, char * argv[]) {
+	signal(SIGINT, _mTestShutdown);
+
+	struct MTestOpts mTestOpts = { 3, -1 };
+	struct mSubParser subparser = {
+		.usage = MTEST_USAGE,
+		.parse = _parseMTestOpts,
+		.extraOptions = MTEST_OPTIONS,
+		.opts = &mTestOpts
+	};
+
+	struct mArguments args;
+	bool parsed = parseArguments(&args, argc, argv, &subparser);
+	if (!args.fname) {
+		parsed = false;
+	}
+	if (!parsed || args.showHelp) {
+		usage(argv[0], MTEST_USAGE);
+		return !parsed;
+	}
+	if (args.showVersion) {
+		version(argv[0]);
+		return 0;
+	}
+	struct mCore* core = mCoreFind(args.fname);
+	if (!core) {
+		return 1;
+	}
+	core->init(core);
+	mCoreInitConfig(core, "mTest");
+	applyArguments(&args, NULL, &core->config);
+
+	mCoreConfigSetDefaultValue(&core->config, "idleOptimization", "remove");
+
+	struct mCoreCallbacks* callbacks = NULL;
+
+#ifdef M_CORE_GBA
+	if (core->platform(core) == mPLATFORM_GBA) {
+		((struct GBA*) core->board)->hardCrash = false;
+
+		_exitSwiImmediate = mTestOpts.exitSwiImmediate;
+		_returnCodeRegister = mTestOpts.returnCodeRegister;
+
+		if (_exitSwiImmediate == 3) {
+			// Hook into SWI 3 (shutdown)
+			callbacks = calloc(1, sizeof(*callbacks));
+			callbacks->context = core;
+			callbacks->shutdown = (void(*)(void*)) _mTestSwi3Callback;
```

Comment:
Any reason this is dynamically allocated? The struct isn't exactly very big.

Reply:
Probably in the mind-set of not consuming memory for GBA callbacks when running a test for GB core (which is dumb, we're hardly memory constrained).
Will remove.